File: matched-substrings-rb/matched_substrings.rb
class Solution
  def match(str1, str2)
    matches = []
    s1 = str1.downcase
    s2 = str2.downcase

# iterate over each substring in the first string
  # ex. given "test":
  #   first outer loop: "t", "te", "tes", "test"
  #   second outer loop: "e", "es", "est"...
# if the 2nd string doesn't include the substring and the substring isn't in the array yet, add it
    (0...s1.length).each do |i|
      (i...s1.length).each do |j|
        substring = s1[i..j]
        if s2.include?(substring) && !matches.include?(substring)
          matches << substring
        end
      end
    end

    return !matches.empty? ? matches : false
  end
end

File: matched-substrings-js/matchedSubstrings.js
// given two strings, return all matching substrings between them (or false if no matches)
// assume case doesn't matter
function matchSubstrings(str1, str2) {
  const substrings = []
  const s1 = str1.toLowerCase()
  const s2 = str2.toLowerCase()

  // iterate over each substring in the 1st string
  // if the 2nd string doesn't include the substring and the substring isn't in the array yet, add it
  for (let i = 0; i < str1.length; i++) {
    for (let j = i + 1; j <= str1.length; j++) {
      let substr = s1.slice(i, j)
      if (s2.includes(substr) && !substrings.includes(substr)) {
        substrings.push(substr)
      }
    }
  }

  return substrings.length !== 0 ? substrings : false
}

module.exports = matchSubstrings
